Add EXECUTORCH_THREADPOOL_SIZE options, default to using only performance cores

GregoryComer opened the pull request:
### Summary
Allow build-time configuration of the thread pool size and default to a performance heuristic.

There are 2 modes that we want to support:
 * Heuristic-based. Choose the number of threads according to a performance heuristic. Use threads equal to the number of detected performance cores, but we can continue to iterate on this by adding fine-grained heuristics for specific chipsets in the future.
 * All cores (threads=cores). This is the current behavior. We need to maintain this as an option for some use cases.

With this PR, the default (for OSS) is to use performance cores. From testing with CV models on ~10 representative devices across the performance spectrum, this gives anywhere from parity with the existing perf to up to a 13x speedup (measured on Pixel 6). Many common devices (S20, S22, iPhone 15 Pro) show a 2-4x speedup.

#### Specifying Threadpool Size

To specify the threadpool size, I've added two preprocessor options (and corresponding CMake options):
* `EXECUTORCH_THREADPOOL_USE_PERFORMANCE_CORES`- Use threads = detected perf cores.
* `EXECUTORCH_THREADPOOL_USE_ALL_LOGICAL_CORES` - Use threads = logical cores.

### Test plan
I've verified that logic functions correctly in OSS by building the executor_runner on M1 Mac and observing the existing logging in cpuinfo_utils.

Measuring MobileNet V3 (exported from examples) on XNNPACK, time to run 100 iterations drops from ~450ms to ~230ms on M1 Pro with this change.

kimishpatel commented on extension/threadpool/threadpool.h:
```diff
@@ -14,6 +14,32 @@
+ *   the threadpool to use 4 threads.
+ */
+
+#ifndef EXECUTORCH_THREADPOOL_SIZE
```
why is this a compile time option? How do you expect users to use it?

GregoryComer replied:
I've piped it through CMake options now. In theory, I'd prefer to have a runtime API, but I don't really see a good way to do this given that it needs to be set prior to threadpool creation, which is implicit. I'm open to ideas. My expectation is that 99.9% of users will use the default perf core option, so needing to override at build time is an acceptable compromise.

kimishpatel reviewed (changes requested):
left a comment

swolchok commented:
I like the 3 mutually exclusive options better. I don't see why we are vague about the "size heuristic" given that we are intentionally making the user select a sizing algorithm; I would just `s/EXECUTORCH_THREADPOOL_USE_SIZE_HEURISTIC/EXECUTORCH_THREADPOOL_USE_PERFORMANCE_CORES/`. Also, it is unclear to me whether `EXECUTORCH_THREADPOOL_USE_ALL_CORES` will count hyperthreading as creating extra cores; might want to call it `EXECUTORCH_THREADPOOL_USE_ALL_PHYSICAL_CORES` or `EXECUTORCH_THREADPOOL_USE_ALL_LOGICAL_CORES` depending on which is correct. (FWIW, not counting hyperthreading would IMO be the correct choice for memory-bandwidth-bound applications.)

Agree with Kimish that it is unclear how users are supposed to correctly choose a fixed size for the thread pool at compile time.

kimishpatel commented on extension/threadpool/CMakeLists.txt:
```diff
@@ -20,6 +20,18 @@ if(NOT CMAKE_CXX_STANDARD)
 endif()
 
+# Threadpool size specifiers. Mutual exclusion is checking in default.cmake.
+set(_threadpool_size_flag)
```
where is this used?

kimishpatel reviewed (commented):
I am not convinced that using `EXECUTORCH_THREADPOOL_SIZE` as build time option is a good. For oss users that are trying to use this option, it feels you are better off with using the unsafe option.

The other possibility is that set_num_thread option will create a new threadpool and use that as a current threadpool without destorying the existing one. 

In any case, I think the other two compile time options look good to me and I am not sure if we have to add `EXECUTORCH_THREADPOOL_SIZE` right now. So the immediate need should be met by the other two, no?

mergennachin commented on extension/threadpool/targets.bzl:
```diff
@@ -34,7 +35,10 @@ def define_common_targets():
             "-DET_USE_THREADPOOL",
-        ],
+        ] + (
+            # Default to perf heuristic (0) in OSS, all cores (-1) otherwise.
```
update this inline comment

kimishpatel reviewed (changes requested):
Left one comment around preserving the default behavior for internal uses

GregoryComer commented on extension/threadpool/threadpool.cpp:
```diff
@@ -101,7 +118,18 @@ ThreadPool* get_threadpool() {
+  // Choose the number of threads according to the EXECUTORCH_THREADPOOL_
+  // options. See the description in threadpool.h.
+
+#if defined(EXECUTORCH_THREADPOOL_USE_ALL_LOGICAL_CORES)
+  // Use threads=cores.
+  static int num_threads = cpuinfo_get_processors_count();
+#else
+  // Set threads equal to the number of performance cores.
+  static int num_threads =
+      ::executorch::extension::cpuinfo::get_num_performant_cores();
+#endif
```
In extension/threadpool/targets.bzl, I changed it to define EXECUTORCH_THREADPOOL_USE_ALL_LOGICAL_CORES when not in OSS, so that should cover this case. If there's a better way to ensure this, I'm definitely open to it. I could add an API to retrieve the threadpool size and add an internal test to verify the behavior, if you'd like.

kimishpatel reviewed (approved):
synced offline for making default behavior chose all core and add cmake flag to chose perf core for oss

GregoryComer committed:
Add EXECUTORCH_THREADPOOL_SIZE options, default to using only performance cores (#14090)

Summary:
Allow build-time configuration of the thread pool size and default to a performance heuristic.

There are 2 modes that we want to support:
 * Heuristic-based. Choose the number of threads according to a performance heuristic. Use threads equal to the number of detected performance cores, but we can continue to iterate on this by adding fine-grained heuristics for specific chipsets in the future.
 * All cores (threads=cores). This is the current behavior. We need to maintain this as an option for some use cases.

With this PR, the default (for OSS) is to use performance cores. From testing with CV models on ~10 representative devices across the performance spectrum, this gives anywhere from parity with the existing perf to up to a 13x speedup (measured on Pixel 6). Many common devices (S20, S22, iPhone 15 Pro) show a 2-4x speedup.

#### Specifying Threadpool Size

To specify the threadpool size, I've added two preprocessor options (and corresponding CMake options):
* `EXECUTORCH_THREADPOOL_USE_PERFORMANCE_CORES`- Use threads = detected perf cores.
* `EXECUTORCH_THREADPOOL_USE_ALL_LOGICAL_CORES` - Use threads = logical cores.


Test Plan:
I've verified that logic functions correctly in OSS by building the executor_runner on M1 Mac and observing the existing logging in cpuinfo_utils.

Measuring MobileNet V3 (exported from examples) on XNNPACK, time to run 100 iterations drops from ~450ms to ~230ms on M1 Pro with this change.

Rollback Plan:

Reviewed By: kimishpatel

Differential Revision: D81965471

Pulled By: GregoryComer
